src/NHibernate/Proxy/FieldInterceptorProxyBuilder.cs

```diff
@@ -0,0 +1,288 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using NHibernate.Intercept;
+using NHibernate.Util;
+
+namespace NHibernate.Proxy
+{
+	internal static class FieldInterceptorProxyBuilder
+	{
+		private const MethodAttributes constructorAttributes =
+			MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.SpecialName | MethodAttributes.RTSpecialName;
+
+		private static readonly System.Type FieldInterceptorAccessorType = typeof(IFieldInterceptorAccessor);
+		private static readonly PropertyInfo AccessorTypeFieldInterceptorProperty =
+			FieldInterceptorAccessorType.GetProperty(nameof(IFieldInterceptorAccessor.FieldInterceptor));
+		private static readonly System.Type FieldInterceptorType = typeof(IFieldInterceptor);
+		private static readonly MethodInfo FieldInterceptorInterceptMethod = FieldInterceptorType.GetMethod(nameof(IFieldInterceptor.Intercept));
+		private static readonly MethodInfo FieldInterceptorMarkDirtyMethod = FieldInterceptorType.GetMethod(nameof(IFieldInterceptor.MarkDirty));
+		private static readonly System.Type AbstractFieldInterceptorType = typeof(AbstractFieldInterceptor);
+		private static readonly FieldInfo AbstractFieldInterceptorInvokeImplementationField =
+			AbstractFieldInterceptorType.GetField(nameof(AbstractFieldInterceptor.InvokeImplementation));
+
+		public static TypeInfo CreateProxyType(System.Type baseType)
+		{
+			// Avoid having a suffix ending with "Proxy", for disambiguation with INHibernateProxy proxies
+			var typeName = $"{baseType.Name}ProxyForFieldInterceptor";
+			var assemblyName = $"{typeName}Assembly";
+			var moduleName = $"{typeName}Module";
+
+			var name = new AssemblyName(assemblyName);
+
+			var assemblyBuilder = ProxyBuilderHelper.DefineDynamicAssembly(AppDomain.CurrentDomain, name);
+			var moduleBuilder = ProxyBuilderHelper.DefineDynamicModule(assemblyBuilder, moduleName);
+
+			const TypeAttributes typeAttributes = TypeAttributes.AutoClass | TypeAttributes.Class | TypeAttributes.Public | TypeAttributes.BeforeFieldInit;
+
+			var interfaces = new HashSet<System.Type>
+			{
+				typeof(IFieldInterceptorAccessor)
+			};
+
+			// Use the object as the base type
+			// since we're not inheriting from any class type
+			var parentType = baseType;
+			if (baseType.IsInterface)
+			{
+				parentType = typeof(object);
+				interfaces.Add(baseType);
```

Comment:
This does indeed not make sens for a field interceptor. The `baseType` can only be an interface for entities mapped as abstract. The field interceptor proxy should never be built for such case, it is built only on entity load from database. And anyway, it can not work in such case, since it uses its base implementation for storing its state.